torch/__init__.py

```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
     Note that deterministic operations tend to have worse performance than
     non-deterministic operations.
 
+    .. note::
+
+        :func:`use_deterministic_algorithms` is not guaranteed to prevent
+        nondeterministic behavior that is caused by an operation performing
+        multiple writes of different data to a single memory location. The
+        order of writes to a single location is not guaranteed, which can cause
+        nondeterminism. This could occur when an output tensor contains internal
+        memory overlaps or if an indexed operation is given duplicate indices.
+        Below are several examples.
+
+    .. code-block:: python
+
+        >>> torch.use_deterministic_algorithms(True)
+
+        # Determinism is not guaranteed for an in-place operation on a tensor
+        # with internal memory overlap
+        >>> a = torch.tensor(1.).as_strided((3,), (0,))
+        >>> a.index_put_((torch.tensor([0, 1, 2]),), torch.tensor([0., 1., 2.]))
+        tensor([2., 2., 2.])
+        >>> a = torch.tensor(1., device='cuda').as_strided((3,), (0,))
+        >>> a.index_put_((torch.tensor([0, 1, 2]).cuda(),), torch.tensor([0., 1., 2.]).cuda())
+        tensor([0., 0., 0.], device='cuda:0')
+
+        # Determinism is not guaranteed for duplicate indices
```

Comment:
@mruberry, since our video call, the plan has changed slightly. If I understand correctly, we should actually guarantee determinism for this case, by alerting without checking for duplicate indices. I can add an alert to `index_put_` if so. `embedding_bag` would also need one. We might have quite a few indexed functions that need an alert.

Reply:
Agreed. Filing issues for those functions would be great; if we can fix them before the PyTorch 1.9 release even better!

Reply:
Actually I just realized I was wrong about `embedding_bag`. It's not nondeterministic, it just has strange behavior if it's given an `offsets` input that doesn't make sense. I'll file an issue for that regardless.